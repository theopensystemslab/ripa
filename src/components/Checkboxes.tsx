import Box from "@material-ui/core/Box";
import Button from "@material-ui/core/Button";
import FormControl from "@material-ui/core/FormControl";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import FormGroup from "@material-ui/core/FormGroup";
import FormLabel from "@material-ui/core/FormLabel";
import { createStyles, makeStyles } from "@material-ui/core/styles";
import { useFormik } from "formik";
import findIndex from "lodash/findIndex";
import sortBy from "lodash/sortBy";
import React, { useState } from "react";
import { FocusWithin } from "react-focus-within";

import Messages from "../shared/components/submit-messages";
import Checkbox from "./Checkbox";
import FocusHandler from "./FocusHandler";
import Question from "./Question";

const useStyles = makeStyles(theme =>
  createStyles({
    formControlLabelRoot: {
      marginLeft: 0,
      marginBottom: theme.spacing(1.5)
    },
    formControlLabel: {
      fontWeight: 400
    }
  })
);

interface ICheckboxes {
  title: string;
  options: object;
  name: string;
  required?: boolean;
  includeSubmit?: boolean;
  handleChange?;
}

const Checkboxes: React.FC<ICheckboxes> = ({
  title = "Title",
  options = {},
  name = "",
  includeSubmit = false,
  handleChange
}) => {
  const [errorMessageVisible, setErrorMessageVisible] = React.useState(false);
  const [successMessageVisible, setSuccessMessageVisible] = React.useState(
    false
  );
  const [submitButtonDisabled, setSubmitButtonDisabled] = React.useState(true);

  const changeHandler = (e, optionId) => {
    let selected = [];

    if (e.target.checked) {
      selected = [...formik.values.selectedOptions, optionId];
    } else {
      selected = formik.values.selectedOptions.filter(el => el !== optionId);
    }

    // sort selected values to match the order of the original options
    selected = sortBy(selected, x =>
      findIndex(Object.keys(options), y => x === y)
    );

    formik.setFieldValue("selectedOptions", selected);

    if (handleChange) handleChange(selected);
  };

  const formik = useFormik({
    initialValues: {
      [name]: "",
      selectedOptions: []
    },
    validate: values => {
      if (values.selectedOptions.length === 0) {
        setErrorMessageVisible(true);
        setSubmitButtonDisabled(true);
      } else {
        setErrorMessageVisible(false);
        setSubmitButtonDisabled(false);
      }
    },
    onSubmit: (values, { resetForm }) => {
      setSuccessMessageVisible(true);
      setTimeout(() => {
        resetForm();
        setSuccessMessageVisible(false);
        setSubmitButtonDisabled(true);
      }, 10000);
    }
  });

  const classes = useStyles();
  return (
    <FocusWithin>
      {({ isFocused, getFocusProps }) => (
        <FocusHandler {...getFocusProps()}>
          <Box py={4}>
            <form
              data-testid="checkboxesComponent"
              onSubmit={formik.handleSubmit}
            >
              <Box pb={1}>
                <Question inFocus={isFocused}>{title}</Question>
              </Box>
              <FormControl component="fieldset">
                <FormLabel
                  component="legend"
                  className={classes.formControlLabelRoot}
                >
                  <Box
                    fontSize="caption.fontSize"
                    color="text.secondary"
                    mb={1}
                  >
                    Select all that apply
                  </Box>
                </FormLabel>
                <Box mb={3}>
                  <FormGroup>
                    {Object.entries(options).map(([optionId, label]) => (
                      <FormControlLabel
                        classes={{
                          root: classes.formControlLabelRoot,
                          label: classes.formControlLabel
                        }}
                        key={optionId}
                        control={
                          <Checkbox
                            checked={formik.values.selectedOptions.includes(
                              optionId
                            )}
                            disableRipple
                            onChange={e => {
                              changeHandler(e, optionId);
                            }}
                            value={optionId}
                            name={optionId}
                          />
                        }
                        label={label}
                      />
                    ))}
                  </FormGroup>
                </Box>
                {errorMessageVisible && formik.touched ? (
                  <Messages
                    type="error"
                    message="Please choose at least one option"
                  />
                ) : null}
                {includeSubmit && (
                  <Button
                    type="submit"
                    disabled={submitButtonDisabled}
                    variant="contained"
                    color="primary"
                  >
                    Save and Continue
                  </Button>
                )}
                {successMessageVisible ? (
                  <Messages
                    type="success"
                    message="Form submitted successfully"
                  />
                ) : null}
              </FormControl>
            </form>
          </Box>
        </FocusHandler>
      )}
    </FocusWithin>
  );
};

export default Checkboxes;
