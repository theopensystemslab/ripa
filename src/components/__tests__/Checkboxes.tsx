import { fireEvent, render } from "@testing-library/react";
import * as React from "react";

import Checkboxes from "../Checkboxes";

describe("Checkboxes Component", () => {
  let title: string;
  let name: string;
  let options: object;
  beforeEach(() => {
    title = "test title";
    name = "test name";
    options = {
      oak: "option 1",
      plane: "option 2",
      redwood: "option 3",
      other: "option 4"
    };
  });
  it("should render items correctly", () => {
    const { getByText } = render(
      <Checkboxes title={title} options={options} name={name} />
    );
    expect(getByText(/test title/i)).toBeInTheDocument();
  });

  it("should render snapshot", () => {
    const { asFragment } = render(
      <Checkboxes title={title} options={options} name={name} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should render success message when selected and enable the submit button", async () => {
    const { getByTestId, getByLabelText, findByText, getByText } = render(
      <Checkboxes
        title={title}
        includeSubmit={true}
        options={options}
        name={name}
      />
    );
    expect(getByText(/save and continue/i).closest("button")).toHaveAttribute(
      "disabled"
    );
    fireEvent.click(getByLabelText(/option 2/i));
    expect(
      getByText(/save and continue/i).closest("button")
    ).not.toHaveAttribute("disabled");
    fireEvent.submit(getByTestId("checkboxesComponent"));
    expect(
      await findByText(/form submitted successfully/i)
    ).toBeInTheDocument();
  });
  it("should render fail message when nothing is selected and disable the submit button", async () => {
    const { getByLabelText, findByText, getByText } = render(
      <Checkboxes
        title={title}
        includeSubmit={true}
        options={options}
        name={name}
      />
    );
    fireEvent.click(getByLabelText(/option 2/i));
    fireEvent.click(getByLabelText(/option 2/i));
    expect(getByText(/save and continue/i).closest("button")).toHaveAttribute(
      "disabled"
    );
    expect(
      await findByText(/please choose at least one option/i)
    ).toBeInTheDocument();
  });
});
