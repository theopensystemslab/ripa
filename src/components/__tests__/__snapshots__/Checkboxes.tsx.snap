// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes Component should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-89"
    tabindex="0"
  >
    <div
      class="MuiBox-root MuiBox-root-91"
    >
      <form
        data-testid="checkboxesComponent"
      >
        <div
          class="MuiBox-root MuiBox-root-92"
        >
          <div
            class="makeStyles-root-93"
          >
            <div
              class="MuiTypography-root makeStyles-question-94 MuiTypography-h4"
            >
              <strong>
                test title
              </strong>
            </div>
          </div>
        </div>
        <fieldset
          class="MuiFormControl-root"
        >
          <legend
            class="MuiFormLabel-root makeStyles-formControlLabelRoot-87"
          >
            <div
              class="MuiBox-root MuiBox-root-138"
            >
              Select all that apply
            </div>
          </legend>
          <div
            class="MuiBox-root MuiBox-root-139"
          >
            <div
              class="MuiFormGroup-root"
            >
              <label
                class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-87"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-157 MuiCheckbox-root makeStyles-checkBoxRoot-148 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-160"
                      data-indeterminate="false"
                      name="oak"
                      type="checkbox"
                      value="oak"
                    />
                    <span
                      class="makeStyles-icon-149"
                      font-size="default"
                    />
                  </span>
                </span>
                <span
                  class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-88 MuiTypography-body1"
                >
                  option 1
                </span>
              </label>
              <label
                class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-87"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-157 MuiCheckbox-root makeStyles-checkBoxRoot-148 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-160"
                      data-indeterminate="false"
                      name="plane"
                      type="checkbox"
                      value="plane"
                    />
                    <span
                      class="makeStyles-icon-149"
                      font-size="default"
                    />
                  </span>
                </span>
                <span
                  class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-88 MuiTypography-body1"
                >
                  option 2
                </span>
              </label>
              <label
                class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-87"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-157 MuiCheckbox-root makeStyles-checkBoxRoot-148 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-160"
                      data-indeterminate="false"
                      name="redwood"
                      type="checkbox"
                      value="redwood"
                    />
                    <span
                      class="makeStyles-icon-149"
                      font-size="default"
                    />
                  </span>
                </span>
                <span
                  class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-88 MuiTypography-body1"
                >
                  option 3
                </span>
              </label>
              <label
                class="MuiFormControlLabel-root makeStyles-formControlLabelRoot-87"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-157 MuiCheckbox-root makeStyles-checkBoxRoot-148 MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-160"
                      data-indeterminate="false"
                      name="other"
                      type="checkbox"
                      value="other"
                    />
                    <span
                      class="makeStyles-icon-149"
                      font-size="default"
                    />
                  </span>
                </span>
                <span
                  class="MuiTypography-root MuiFormControlLabel-label makeStyles-formControlLabel-88 MuiTypography-body1"
                >
                  option 4
                </span>
              </label>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
